### PR TITLE
trait alias infrastructure

### PR DESCRIPTION
**Pull request by durka**
This will be an implementation of trait aliases (RFC 1733, #41517).

Progress so far:

- [x] Feature gate
- [x] Add to parser
  - [x] `where` clauses
    - [x] prohibit LHS type parameter bounds via AST validation https://github.com/rust-lang/rust/pull/45047#discussion_r143575575
- [x] Add to AST and HIR
  - [x] make a separate PathSource for trait alias contexts https://github.com/rust-lang/rust/pull/45047#discussion_r143353932
- [x] Stub out enough of typeck and resolve to just barely not ICE

Postponed:

- [ ] Actually implement the alias part
- [ ] #21903
- [ ] #24010

I need some pointers on where to start with that last one. The test currently does this:

```
error[E0283]: type annotations required: cannot resolve `_: CD`
  --> src/test/run-pass/trait-alias.rs:34:16
   |
34 |     let both = foo();
   |                ^^^
   |
   = note: required by `foo`
```

### CODE REVIEW

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -391,6 +391,9 @@ impl<'a> Resolver<'a> {
                 self.define(parent, ident, TypeNS, (module, vis, sp, expansion));
                 self.current_module = module;
             }
+            ItemKind::TraitAlias(..) => {
+                // TODO: do we need to do anything here?
```
Comment (petrochenkov) [inline src/librustc_resolve/build_reduced_graph.rs]: Yes, the same thing as in `ItemKind::Ty` (only with `Def::TraitAlias`).

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -5002,21 +5002,30 @@ impl<'a> Parser<'a> {
...
+                            self.recover_stmt_(SemiColonMode::Break, BlockMode::Break);
+                        }
                     }
                 }
             }
+            Ok((ident, ItemKind::Trait(unsafety, tps, bounds, trait_items), None))
+        } else {
+            self.expect(&token::Eq)?;
+            let bounds = self.parse_ty_param_bounds()?;
+            self.expect(&token::Semi)?;
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: `where` clauses?

File: src/libsyntax/ast.rs [hunk 3]
```diff
@@ -1927,6 +1927,10 @@ pub enum ItemKind {
     ///
     /// E.g. `trait Foo { .. }` or `trait Foo<T> { .. }`
     Trait(Unsafety, Generics, TyParamBounds, Vec<TraitItem>),
+    /// Trait alias
+    /// 
+    /// E.g. `trait Foo = Bar + Quux;`
+    TraitAlias(Generics, TyParamBounds),
```
Comment (petrochenkov) [inline src/libsyntax/ast.rs]: I need to reread the RFC, but IIRC there should be two sets of predicates - one for the alias itself (LHS) and one for the alias substitution result.
I.e.
```rust
trait Alias<T: Clone> = where T: Default;

=>

trait Alias<T> whereLHS T: Clone = whereRHS T: Default;
```
(But maybe they are equivalent?)

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustc_resolve/build_reduced_graph.rs]: You'll also have to add a new variant to `PathSource` for contexts permitting trait alias paths (they are slightly different than contexts permitting trait paths.)

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (petrochenkov) [reply src/libsyntax/ast.rs]: I mean `Alias<NoClone>` is an immediate error, but `Alias<CloneNoDefault>` is a valid always-false predicate.

**Comment by petrochenkov** [comment]
@durka 
>I need some pointers on where to start with that last one. 

TLDR is "look how `instantiate_poly_trait_ref` and friends (in `predicates_of`) turn `hir::` bounds and `where` clauses into `ty::Predicate`s and do the same thing".

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (arielb1) [reply src/libsyntax/ast.rs]: I think they are equivalent  - both are added as predicates.

Comment (petrochenkov) [reply src/libsyntax/ast.rs]: This should be related to https://github.com/rust-lang/rust/issues/21903 somehow.

Comment (durka) [reply src/libsyntax/ast.rs]: The RFC only mentions `whereRHS`.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (durka) [reply src/librustc_resolve/build_reduced_graph.rs]: Why would the contexts be different? Because `impl Alias {}` won't be allowed?

Comment (petrochenkov) [reply src/librustc_resolve/build_reduced_graph.rs]: Yes.
So you'll automatically get "expected a trait, found a trait alias" errors on `impl Alias {}`.

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (petrochenkov) [reply src/libsyntax/ast.rs]: @durka 
Could you prohibit bounds on trait alias' generic parameters in `ast_validation` for now?
```
trait TraitAlias<T: Bound /* <-- error, this one represents "whereLHS" */>
```

Comment (durka) [reply src/libsyntax/ast.rs]: Sounds good.

**Comment by durka** [comment]
I'm in way over my head here.

I [stopped `explicit_predicates_of`](https://github.com/rust-lang/rust/pull/45047/files#diff-4ed25c00aceb84666fca639cf8101c7cR1444) from adding a `Self: TheAlias` bound when evaluating `TheAlias` directly. But that doesn't work when evaluating something _using_ the alias, like `fn foo<T: TheAlias>(...) { ... }`. So I added a [hack](https://github.com/rust-lang/rust/pull/45047/files#diff-1b4d50b82a87f1031cd5f99b64123de7R1450) to `Bounds::predicates` that scans for aliases and calls `super_predicates_of` to eliminate them. It seems really fishy, but it only gets me to another error, shown here with logs:

```
DEBUG:rustc_typeck::check::wfcheck: check_item_well_formed(it.id=73, it.name=foo)
DEBUG:rustc::traits::fulfill: select(obligation-forest-size=0)
DEBUG:rustc::traits::fulfill: select: starting another iteration
DEBUG:rustc::traits::fulfill: select: outcome=Outcome { completed: [], errors: [], stalled: true }
DEBUG:rustc::traits::fulfill: select(0 predicates remaining, 0 errors) done
DEBUG:rustc_typeck::check: normalize_associated_types_in(value=Binder(([]; variadic: false)->(T, T)))
DEBUG:rustc_typeck::check: normalize_associated_types_in: result=Binder(([]; variadic: false)->(T, T)) predicates=[]
DEBUG:rustc_typeck::check: normalize_associated_types_in(value=InstantiatedPredicates([Binder(TraitPredicate(<T as std::marker::Sized>)), Binder(TraitPredicate(<Self as std::clone::Clone>)), Binder(TraitPredicate(<Self as std::default::Default>))]))
DEBUG:rustc_typeck::check: normalize_associated_types_in: result=InstantiatedPredicates([Binder(TraitPredicate(<T as std::marker::Sized>)), Binder(TraitPredicate(<Self as std::clone::Clone>)), Binder(TraitPredicate(<Self as std::default::Default>))]) predicates=[]
DEBUG:rustc_typeck::check::wfcheck: check_item_fn: predicates=InstantiatedPredicates([Binder(TraitPredicate(<T as std::marker::Sized>)), Binder(TraitPredicate(<Self as std::clone::Clone>)), Binder(TraitPredicate(<Self as std::default::Default>))])
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 1
DEBUG:rustc_typeck::check: normalize_associated_types_in(value=Binder(([]; variadic: false)->(T, T)))
DEBUG:rustc_typeck::check: normalize_associated_types_in: result=Binder(([]; variadic: false)->(T, T)) predicates=[]
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 2
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 3
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 4
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 5
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=WF((T, T)),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=WF((T, T)),depth=0))
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 6
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 7
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=0))
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=WF(T),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=WF(T),depth=0))
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0))
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=Binder(TraitPredicate(<Self as std::clone::Clone>)),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=Binder(TraitPredicate(<Self as std::clone::Clone>)),depth=0))
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=WF(Self),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=WF(Self),depth=0))
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0))
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=Binder(TraitPredicate(<Self as std::default::Default>)),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=Binder(TraitPredicate(<Self as std::default::Default>)),depth=0))
DEBUG:rustc_typeck::check: register_predicate(Obligation(predicate=WF(Self),depth=0))
DEBUG:rustc::traits::fulfill: register_predicate_obligation(obligation=Obligation(predicate=WF(Self),depth=0))
DEBUG:rustc_typeck::check::wfcheck: check_item_fn_or_method: 8
DEBUG:rustc_typeck::check: select_all_obligations_or_error
DEBUG:rustc::traits::fulfill: select(obligation-forest-size=7)
DEBUG:rustc::traits::fulfill: select: starting another iteration
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<T as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<Self as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<Self as std::clone::Clone>))` at depth 0 yielded Ok(Some)
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<Self as std::default::Default>))` at depth 0 yielded Ok(Some)
DEBUG:rustc::traits::fulfill: select: outcome=Outcome { completed: [PendingPredicateObligation { obligation: Obligation(predicate=Binder(TraitPredicate(<Self as std::default::Default>)),depth=0), stalled_on: [] }, PendingPredicateObligation { obligation: Obligation(predicate=WF(Self),depth=0), stalled_on: [] }, PendingPredicateObligation { obligation: Obligation(predicate=Binder(TraitPredicate(<Self as std::clone::Clone>)),depth=0), stalled_on: [] }, PendingPredicateObligation { obligation: Obligation(predicate=WF(T),depth=0), stalled_on: [] }], errors: [], stalled: false }
DEBUG:rustc::traits::fulfill: select: starting another iteration
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<T as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<Self as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: select: outcome=Outcome { completed: [], errors: [], stalled: true }
DEBUG:rustc::traits::fulfill: select(3 predicates remaining, 0 errors) done
DEBUG:rustc::traits::fulfill: select(obligation-forest-size=3)
DEBUG:rustc::traits::fulfill: select: starting another iteration
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<T as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<Self as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: select: outcome=Outcome { completed: [], errors: [], stalled: true }
DEBUG:rustc::traits::fulfill: select(3 predicates remaining, 0 errors) done
DEBUG:rustc::traits::fulfill: select(obligation-forest-size=3)
DEBUG:rustc::traits::fulfill: select: starting another iteration
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<T as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: selecting trait `Binder(TraitPredicate(<Self as std::marker::Sized>))` at depth 0 yielded Ok(None)
DEBUG:rustc::traits::fulfill: process_predicate: pending obligation Obligation(predicate=Binder(TraitPredicate(<Self as std::marker::Sized>)),depth=0) now stalled on []
DEBUG:rustc::traits::fulfill: select: outcome=Outcome { completed: [], errors: [], stalled: true }
DEBUG:rustc::traits::fulfill: select(3 predicates remaining, 0 errors) done
error[E0282]: type annotations needed
  --> src/test/run-pass/trait-alias.rs:27:1
   |
27 | / fn foo<T: CD>() -> (T, T) {
28 | |     let one = T::default();
29 | |     let two = one.clone();
30 | |     (one, two)
31 | | }
   | |_^ cannot infer type for `T`
```

I can't see where the error is coming from, since trait selection keeps saying no errors and the log looks similar to successful fulfillments, to my eye. Where should I look next?

File: src/librustc/hir/print.rs [hunk 4]
```diff
@@ -735,6 +735,27 @@ impl<'a> State<'a> {
...
                 self.bclose(item.span)?;
             }
+            hir::ItemTraitAlias(ref generics, ref bounds) => {
+                self.head("")?;
+                self.print_visibility(&item.vis)?;
+                self.word_nbsp("trait")?;
+                self.print_name(item.name)?;
+                self.print_generics(generics)?;
+                let mut real_bounds = Vec::with_capacity(bounds.len());
+                // FIXME(durka) this seems to be some quite outdated syntax
```
Comment (nikomatsakis) [inline src/librustc/hir/print.rs]: yeah, you used to have to say `trait Foo for ?Sized`, but now that's the default (and you have to write `trait Foo: Sized`)

Comment (nikomatsakis) [reply src/librustc/hir/print.rs]: I would assume that `TraitBoundModifier::Maybe` in this context is illegal, right? i.e., `trait Foo = ?Sized` is kind of nonsense.

File: src/librustc_typeck/astconv.rs [hunk 5]
```diff
@@ -1446,7 +1447,15 @@ impl<'a, 'gcx, 'tcx> Bounds<'tcx> {
         }
 
         for bound_trait_ref in &self.trait_bounds {
-            vec.push(bound_trait_ref.to_predicate());
+            if let Some(node_id) = tcx.hir.as_local_node_id(bound_trait_ref.def_id()) {
+                if let hir::ItemTraitAlias(..) = tcx.hir.expect_item(node_id).node {
```
Comment (nikomatsakis) [inline src/librustc_typeck/astconv.rs]: Hmm. This doesn't seem like the right place to be doing these sorts of tests. By the time we get to `ty::TraitRef`, aliases should already be expanded, I think -- or, alternatively, we should do the expansion during trait resolution. The latter seems better and more like the strategy we want long term (not sure if that's the direction you've been going, still reading into the PR). Either way, this bit of code should't have to change.

File: src/librustc_typeck/collect.rs [hunk 6]
```diff
@@ -668,7 +668,7 @@ fn adt_def<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 /// Ensures that the super-predicates of the trait with def-id
 /// trait_def_id are converted and stored. This also ensures that
 /// the transitive super-predicates are converted;
-fn super_predicates_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+pub fn super_predicates_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```
Comment (nikomatsakis) [inline src/librustc_typeck/collect.rs]: Why is this public now? It's a provider so I think it should not have to be, we should be using the query instead.

**Review by nikomatsakis — COMMENTED** [review]
OK, did a quick read through, here are two off-the-cuff comments.

**Comment by nikomatsakis** [comment]
Question:

Is the overall strategy here to treat a "trait alias" `trait Foo = Bar` as its own kind of trait, but:

- it has 'supertraits' `Bar`, hence we elaborate `T: Foo` into `T: Bar`?
- to prove that `T: Foo` is true, we must prove that `T: Bar` is true?

If so, I approve. =) This is how I had hoped to do it, though for some reason I thought we might need to do some kind of "syntactic expansion" first. But now I don't see a reason we have to go through that step.

File: src/librustc_typeck/collect.rs [hunk 7]
```diff
@@ -1420,14 +1436,20 @@ fn explicit_predicates_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
...
 
-        // Add in a predicate that `Self:Trait` (where `Trait` is the
-        // current trait).  This is needed for builtin bounds.
-        predicates.push(trait_ref.to_poly_trait_ref().to_predicate());
+        debug!("explicit_predicates_of: 5 predicates={:?}", predicates);
+
+        if items.is_some() {
+            // Add in a predicate that `Self:Trait` (where `Trait` is the
+            // current trait).  This is needed for builtin bounds.
+            predicates.push(trait_ref.to_poly_trait_ref().to_predicate());
```
Comment (nikomatsakis) [inline src/librustc_typeck/collect.rs]: I feel like we should be able to treat traits and trait aliases the same here if we set things up right (i.e., either both or neither would get this extra predicate). But it's not a big deal I don't think.

**Comment by nikomatsakis** [comment]
@durka 

> But that doesn't work when evaluating something _using_ the alias, like `fn foo<T: TheAlias>(...) { ... }`, because `astconv` simply constructs the predicates itself without consulting `collect`.

I think that [this comment is exactly your problem here](https://github.com/rust-lang/rust/pull/45047/files#r144963205), though it's possible I misunderstand. I think that hunk should basically be reverted. Instead, we want to extend the `select` logic to understand trait aliases.

There are ways I can see to do this. One way would be that, in HIR lowering, we would basically generate a synthetic impl like `impl<..., T> TraitAlias for T`. Effectively something like `trait Foo = Bar` would be [desugared to something like this](https://play.rust-lang.org/?gist=71edb1881e7d3aae5fb3bd8e1ee4a120&version=undefined). This is kind of slick but will require some effort to get the error messages nice, of course.

Some questions though:

- Can trait aliases be used with object types? I believe so, and that seems to imply we have to be careful about how we are handling supertraits here, since that influences object safety and in particular what methods can be invoked.
- I think what we probably want is a third category:
    - Right now, we have *superpredicates* (which are basically `Self: Foo` where-clauses) and *predicates*.
    - Superpredicates are used to construct trait objects and also used in elaboration.
    - I suspect we want to separate those two roles, so that trait aliases can have one set of superpredicates that are used for trait objects, and another that are used for elaboration (which includes **all** where-clauses on the trait alias).

The other way would be to extend trait selection to treat "trait aliases" as a kind of native concept. This isn't *too much* work, we probably have to add a [new vtable variant](https://github.com/rust-lang/rust/blob/master/src/librustc/traits/mod.rs#L287) like `TraitAlias`:

```rust
pub enum Vtable<'tcx, N> {
    ...
    VtableAlias(VtableAliasData),
    ...
}

pub struct VtableAliasData<N> {
    pub alias_def_id: DefId,
    pub substs: &'tcx Substs<'tcx>,
    pub nested: Vec<N>,
}
```

and then extend [this logic here](https://github.com/rust-lang/rust/blob/master/src/librustc/traits/select.rs#L1334-L1367) to add an "alias candidate" for aliases and things. This promises to be straightforward but a lot of boilerplate.

Personally, I'm inclined to go with the "lower at HIR construction time" route myself.

File: src/librustc_typeck/collect.rs (hunk 6 above)
Comment (durka) [reply src/librustc_typeck/collect.rs]: It's public because I needed it in the hack that I added to astconv, which you correctly pointed out just above is the wrong place 😛

File: src/librustc_typeck/collect.rs (hunk 7 above)
Comment (durka) [reply src/librustc_typeck/collect.rs]: Yeah, this is the part that works if you say "what are the preds of `TraitAlias`" (i.e. it adds only the super predicates and not `TraitAlias` itself) but falls over if you ask "what are the preds of <some item that refers to `TraitAlias`>", which may indicate it's too early or late to be distinguishing traits from aliases.

**Comment by durka** [comment]
What you asked about was indeed my plan, and I figured I would get to the trait object subleties once I got everything to stop ICEing all the time 😛.

About the HIR desugaring: hmm, but I [claimed that sugar was accurate](https://github.com/rust-lang/rfcs/pull/1733#issuecomment-243840014) early in the RFC thread, I was corrected. The reasons I was wrong were that (1) it [wouldn't allow you to impl a trait alias](https://github.com/rust-lang/rfcs/pull/1733#issuecomment-244005410) (which we decided not to allow in the end, but could be relitigated in the future), and (2) it means that `let x: Box<AliasOfBar> = Box::new(&foo as &Bar);` [would not work](https://github.com/rust-lang/rfcs/pull/1733#issuecomment-245012707). So these sort of seem to argue for doing the expansion deeper in the compiler.

File: src/librustc/hir/print.rs (hunk 4 above)
Comment (durka) [reply src/librustc/hir/print.rs]: Agreed.

File: src/librustc_typeck/astconv.rs (hunk 5 above)
Comment (durka) [reply src/librustc_typeck/astconv.rs]: Agreed (see below).

**Comment by nikomatsakis** [comment]
@durka 

>  The reasons I was wrong were

Also the supertraits point that I raised earlier =)

> it wouldn't allow you to impl a trait alias 

While true, I think this isn't quite a blocker. In particular, if we do decide to permit that, I feel like it would mostly mean that we want to adjust the desugaring further -- i.e., maybe we desugar **that** impl by unrolling the trait alias one step (in fact, that's what I would probably expect us to do), but leaving the "magic, implicit" impl for the alias in place.

(Really, I want to rework trait selection so that everything winds up getting lowered internally to simpler primitives anyway, which would kind of render this discussion semi-moot, but we're not there yet.)

> it means that `let x: Box<AliasOfBar> = Box::new(foo) as Box<Bar>;` would not work.

This is an interesting point, but I think orthogonal from the question of whether to desugar into a "synthetic HIR impl" vs customizing trait selection. It might argue for another approach entirely, though I suspect it can be made to work.

Basically, the problem here is that `Box<dyn AliasOfBar>` would be lowered to a trait object whose trait def-id is **the alias**, and we don't (right now) permit much in the way of interconversion between `dyn` types.  Note that I'm adopting the `dyn` syntax to make the "trait used in type" vs "trait used as logical predicate" distinction more evident, since it is crucial here. Basically, either adding the impl or hacking resolve solves the "trait used as logical predicate" problem (I believe), but neither solves the "trait used in type" scenario.

I think we can correct this in two ways:

- When creating a trait object type, we "dereference" trait aliases -- i.e., `dyn AliasOfBar`, in the compiler, would be unrolled into `dyn Bar`. Now the mismatch goes away (but the user's error messages will show `dyn Bar` instead of `dyn AliasOfBar`).
- We could hack up trait object unification so that unifying `dyn T` and `dyn U` succeeds if `T` is an alias for `U` or vice versa. This sounds like a pain in the neck.

I favor the first idea. =)

**Comment by durka** [comment]
First idea sounds good to me! Type errors already see through type aliases, so the error message doesn't seem like a blocker to me. But, does it present a problem with more complex aliases? For example, `Box<dyn Clone + Default where for<'a> Self: Foo<'a>>` doesn't quite parse :)

**Comment by durka** [comment]
@nikomatsakis So, basically I should revert most of my changes to `collect` and `astconv` and focus on `select` instead? Can you give some more specific pointers?

**Comment by nikomatsakis** [comment]
@durka 

> For example, `Box<dyn Clone + Default where for<'a> Self: Foo<'a>>` doesn't quite parse :)

First off, I would say it's a good idea to produce a first PR that basically just makes it an error to use trait aliases in an object type. 

In any case, I imagine that after we expand the trait alias we'll have to examine the traits and make sure that they fit what our current set of object types can support.

You can already type things (e.g., `dyn (Display + Debug)`) that are not legal object types at the moment.

**Comment by petrochenkov** [comment]
I've read the previous messages, but I still don't understand why trait aliases can't be expanded in astconv/collect in the same way like type aliases are expanded (doesn't matter where, in trait objects or not in trait objects).

**Comment by durka** [comment]
Where are type aliases expanded?

**Comment by petrochenkov** [comment]
@durka 
>Where are type aliases expanded?

`fn def_to_ty`/`fn ast_path_to_ty`

**Comment by carols10cents** [comment]
Triage ping to keep this on your radar @durka!

**Comment by durka** [comment]
It's still on my radar!

**Comment by nikomatsakis** [comment]
@petrochenkov 

> I've read the previous messages, but I still don't understand why trait aliases can't be expanded in astconv/collect in the same exact way like type aliases are expanded (doesn't matter where, in trait objects or not in trait objects).

They could, but I'd prefer to avoid that if we can. For one thing, it does not offer an easy path to handling where clauses. I actually hope to refactor so that type aliases are also expanded in the trait system eventually, as a kind of poor man's associated type.

**Comment by nikomatsakis** [comment]
@durka how's it going? Still have time to work on this?

**Comment by durka** [comment]
Not until next week unfortunately :(

**Comment by petrochenkov** [comment]
@nikomatsakis 
>They could, but I'd prefer to avoid that if we can. For one thing, it does not offer an easy path to handling where clauses. I actually hope to refactor so that type aliases are also expanded in the trait system eventually, as a kind of poor man's associated type.

My only concern is implementation complexity (for someone without good knowledge of middle-end (traits, type checking, etc)). Frontend is more straightforward and intuitive after all.
Given that type aliases are still not moved into type checker for long time despite [issues](https://github.com/rust-lang/rust/issues/21903), my conclusion was that it's better to have initial implementation of trait aliases in the frontend + refactoring some time in the future, than no implementation of trait aliases.

**Comment by nikomatsakis** [comment]
@petrochenkov as long as they are feature-gated, I suppose the risk is minimal, though I think it's probably not too hard to do it. I have to review the comments now to remember what I had in mind =)

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (durka) [reply src/libsyntax/ast.rs]: Added an error for this and also `trait Alias<T = ()>`. [cfail](https://github.com/rust-lang/rust/pull/45047/files#diff-c764b8ef26325201c0a34cadf64a856a)

**Comment by petrochenkov** [comment]
It may be reasonable to split the implementation into >1 PRs like it was done with few other implementation period tasks (e.g. https://github.com/rust-lang/rust/pull/45904).
The first PR contains data structures, visitors, parsing, name resolution, etc. - all the boilerplate. All uses of trait aliases are lowered into *nothing* (no predicates) in astconv/collect after reporting an "unimplemented" error.
The second PR would then implement the more compact "essential" part, less susceptible to bit rotting etc.

**Comment by durka** [comment]
Yeah that's probably a good idea. Sorry for being slow on this.

**Comment by durka** [comment]
OK, I rebased and retooled the PR as @petrochenkov suggested. Now it contains only AST/HIR scaffolding for trait aliases, and you get an error if you try to use them.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (durka) [reply src/librustc_resolve/build_reduced_graph.rs]: This already happens with no modifications to `PathSource`.

Comment (durka) [reply src/librustc_resolve/build_reduced_graph.rs]: Oh but maybe this is needed to check for `impl Alias for Struct {}`?

**Comment by petrochenkov** [comment]
@durka 
>Oh but maybe this is needed to check for impl Alias for Struct {}?

Yes, `impl Trait for Type { ... }` (and also soon to be removed `impl Trait for .. { ... }`) should accept traits, but not trait aliases.

**Comment by petrochenkov** [comment]
Also, trait alias items look pretty much implemented, so it's not necessary to error on their definition, only on use.

**Review by nikomatsakis — APPROVED** [review]
Looks good.

**Comment by nikomatsakis** [comment]
I'd be ok w/ landing this as is, or after making the alterations that @petrochenkov suggested (erroring only on use).

**Comment by durka** [comment]
I'm not quite sure what is meant by erroring on definition but not use. I'm working on the `PathSource` thing, right now I don't see *any* error on `impl Alias for Struct {}` which is weird.

**Comment by petrochenkov** [comment]
@durka 
>I'm not quite sure what is meant by erroring on definition but not use

The other way around, use but not definition.
```rust
trait Alias = MyTrait; // OK, definition

fn f<T: Alias>() {} // ERROR, use
```

**Comment by durka** [comment]
I think I have the `PathSource` stuff working. It was a little tricky because most of the code doesn't care whether a trait is an alias or not.

File: src/librustc_resolve/lib.rs [hunk 8]
```diff
@@ -390,12 +390,18 @@ impl PatternSource {
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+enum AliasPossibility {
```
Comment (durka) [inline src/librustc_resolve/lib.rs]: I'll be the first to admit this is a bad name.

File: src/librustc_resolve/lib.rs [hunk 9]
```diff
@@ -2076,7 +2098,7 @@ impl<'a> Resolver<'a> {
                                                        &path,
                                                        trait_ref.path.span,
                                                        trait_ref.path.segments.last().unwrap().span,
-                                                       PathSource::Trait)
+                                                       PathSource::Trait(AliasPossibility::No))
```
Comment (durka) [inline src/librustc_resolve/lib.rs]: This function `with_optional_trait_ref` is only invoked when resolving a trait impl or auto trait impl. But there's no documentation so I don't know whether that's intended or not. If not, it's the wrong place to make this distinction.

**Comment by petrochenkov** [comment]
Excellent, let's land this already.
@bors r+

File: src/librustc_resolve/lib.rs (hunk 9 above)
Comment (durka) [reply src/librustc_resolve/lib.rs]: OK, I guess this was tacitly endorsed, so I'll add comments in a future commit.

**Comment by durka** [comment]
rustfmt broke. I thought toolstate was automatic now?

**Comment by petrochenkov** [comment]
@durka 
Not yet (https://github.com/rust-lang/rust/pull/46554).

@bors delegate+

**Comment by durka** [comment]
Same failure.

**Comment by kennytm** [comment]
Whenever you break `rustfmt`, you need to break `rls` too.

**Comment by durka** [comment]
Clippy doesn't build locally either. But it seems to be an unrelated error. Maybe it's somehow out of sync with submodules?

**Comment by kennytm** [comment]
@durka Yeah clippy is an unrelated error. You may ignore it for now.
